js/web/lib/wasm/jsep/webgpu/ops/common.ts

```diff
@@ -15,102 +16,554 @@ import {ShapeUtil} from '../../util';
-      shape.length < 2 ? `${varIndices}=${varOffset};` : `ih_o2i_${name}(${varOffset}, &${varIndices});`;
-
-  const offsets: string[] = [];
-  if (shape.length === 0) {
-    offsets.push('0u');
-  } else if (shape.length < 2) {
-    offsets.push('(*indices)');
-  } else {
-    for (let i = shape.length - 1; i >= 0; i--) {
-      offsets.push(`${strides[i]}u * ((*indices)[${i}])`);
-    }
-  }
+      const offsetToIndices = (varOffset: string) => rank < 2 ? varOffset : `o2i_${name}(${varOffset})`;
+
+      const offsets: string[] = [];
+      if (rank >= 2) {
+        for (let i = rank - 1; i >= 0; i--) {
+          offsets.push(`${strides[i]}u * (indices[${i}])`);
+        }
+      }
 
-  const i2oImpl = shape.length < 2 ? '' : `
-  fn ih_i2o_${name}(indices: ptr<function, ${iType}>) -> u32 {
+      const indicesToOffsetImplementation = rank < 2 ? '' : `
+  fn i2o_${name}(indices: ${type.indices}) -> u32 {
     return ${offsets.join('+')};
   }`;
 
-  const i2oExpression = (varIndices: string, isPtr?: boolean) =>
-      shape.length < 2 ? `(${isPtr ? '*' : ''}${varIndices})` : `ih_i2o_${name}(${isPtr ? '' : '&'}${varIndices})`;
+      const indicesToOffset = (varIndices: string) => rank < 2 ? varIndices : `i2o_${name}(${varIndices})`;
 
-  const indicesVariableDeclaration = (v: string, init?: string[]) =>
-      `var ${v}:${iType}${init ? `=${iType}(${init.join(',')})` : ''};`;
+      const indices = (...init: ReadonlyArray<number|string>) =>
+          rank === 0 ? '0u' : `${type.indices}(${init.map(normalizeDim).join(',')})`;
 
-  return {o2iImpl, o2iCall, i2oImpl, i2oExpression, indicesVariableDeclaration, iType};
-};
+      const indicesGet = (varIndices: string, idx: number|string) => {
+        if (rank < 2) {
+          return `${varIndices}`;
+        } else {
+          return `${varIndices}[${idx}]`;
+        }
+      };
+
+      const indicesSet = (varIndices: string, idx: number|string, value: string) => {
+        if (rank < 2) {
+          return `${varIndices}=${value};`;
+        } else {
+          return `${varIndices}[${idx}]=${value};`;
+        }
+      };
+
+      const setByOffset = (offset: number|string, value: string) => (() => {
+        if (type.storage === type.value) {
+          return `${name}[${offset}]=${value};`;
+        } else if (type.storage === 'vec2<u32>' && type.value === 'i32') {
+          // int64, components === 1
+          return `${name}[${offset}]=vec2<u32>(u32(${value}), select(0u, 0xFFFFFFFFu, ${value} < 0));`;
+        } else if (type.storage === 'vec2<u32>' && type.value === 'u32') {
+          // uint64, components === 1
+          return `${name}[${offset}]=vec2<u32>(u32(${value}), 0u);`;
+        } else if (type.storage === 'u32' && type.value === 'vec4<bool>') {
+          // bool, components === 4
+          return `${name}[${offset}]=u32(select(0u,0x1u,${value}.x)|select(0u,0x100u,${value}.y)|select(0u,0x10000u,${
+              value}.z)|select(0u,0x1000000u,${value}.w));`;
```

Comment:
Just fyi. My colleague just told me that `dot(vec4<u32>(0x1, 0x100, 0x10000, 0x1000000), vec4<u32>(${value}))` may be better than `u32(select(0u,0x1u,${value}.x)|select(0u,0x100u,${value}.y)|select(0u,0x10000u,${value}.z)|select(0u,0x1000000u,${value}.w))` since it can avoid the conditional judgment. It's ok if you keep the current one.